### PR TITLE
sessionctx, variable: add skip name resolve

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/9778

Problem Summary:

I was investigating linter issues, when I noticed that the silencing of the error in `getHostByIP` might not be completely correct. If the reverse lookup fails (`net.LookupAddr`) - is the original IP returned? I expect this behavior is undefined, so authentication might fail when it doesn't need to.

Then I was wondering why there is no method to skip name resolution like there is in MySQL, since it is a common problem in large scale systems. There has been a feature request @ https://github.com/pingcap/tidb/issues/9778

In MySQL skip-name-resolve is non-dynamic, but this does not make so much sense for TiDB, so I've changed it to GLOBAL-only scope.

I also don't think it is correct to silently drop errors from name resolution (although the fact that they were previously noisy and this was fixed in https://github.com/pingcap/tidb/pull/13989 indicates that skip-name resolve as a feature has merit). My resolution was to reintroduce an error message, but switch it to debug level.

### What is changed and how it works?

What's Changed:

Support has been added for skip-name-resolve.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

Side effects

- Increased code complexity
 
### Release note 

- Support has been added for skip-name-resolve.

### CODE REVIEW

**Review by xhebox — APPROVED** [review]
LGTM

File: session/session.go [hunk 1]
```diff
@@ -2228,14 +2228,24 @@ func (s *session) AuthWithoutVerification(user *auth.UserIdentity) bool {
 	return false
 }
 
-func getHostByIP(ip string) []string {
+func (s *session) getHostByIP(ip string) []string {
 	if ip == "127.0.0.1" {
 		return []string{variable.DefHostname}
 	}
+	skipNameResolve, err := s.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.SkipNameResolve)
+	if err == nil && variable.TiDBOptOn(skipNameResolve) {
```
Comment (djshow832) [inline session/session.go:2236]: Are you intended to continue even when `err != nil`?

Comment (morgo) [reply session/session.go:2236]: Yes. The option defaults to `OFF` (and always will). It's only if can be successfully read as `ON` that we take the short-path and return the IP address. See: https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_skip_name_resolve

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**